### PR TITLE
Add DDExtrudedPolygon, DDPolycone, and DDPolyhedra shapes for reading shape parameters

### PR DESCRIPTION
**Pull request by cvuosalo**
The `DetectorDescription/DDCMS/interface/DDShapes.h` file has provided classes for reading several shapes: `DDBox`, `DDCons`, `DDTrap`, `DDTubs`, and `DDTruncTubs`. This PR adds `DDExtrudedPolygon`, `DDPolycone`, and `DDPolyhedra`. These new classes are implemented in a new way that avoids any data members.

After these new classes are put to use, if it becomes clear that their implementation is superior to the implementation of the older shapes, a follow-up PR can be done to re-implement the older shapes in the new way.

Note that with the new shapes, the calling code must ensure that the matching type of solid is passed to the constructor; otherwise, a run-time exception could occur when a shape method is called.

#### PR validation:

A new unit test is added that exercises each of the shapes. For each kind of shape, a detailed comparison is made to reference values to ensure that the parameters are being read out correctly. In this PR, some of the reference XML files were slightly revised to eliminate degenerate parameter values and facilitate robust testing.

### CODE REVIEW

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
@ghugo83 @wpcarvalho FYI

**Review by ianna — COMMENTED** [review]
@cvuosalo - this implementation is pretty generic. Why it is submitted to CMSSW and not to DD4hep repo?
Also, this cannot be used to create shapes in C++ Algorithms. Do you plan to extend this? This would mean reimplementing the DD.

**Comment by cvuosalo** [comment]
@ianna This PR is submitted to CMSSW because it is not ready to be submitted to the DD4hep repo. We should get experience with using these classes in CMSSW to see if there are issues or ways to improve the interfaces. Then the DDShapes implementation should be made more consistent and polished.
The DD4hep user base is much wider than that for CMSSW. Use cases for DDShapes in the DD4hep realm would need to be evaluated. Additional shapes like DDTrd1 that are of dubious usefulness in CMSSW might need to be added for DD4hep. Then the consistency of user interface design across DD4hep for shapes would also need to be considered. I think there would be a lot of work involved to prepare and submit these DDShapes to the DD4hep repo.

Concerning creating shapes, these classes are intended as read-only, not for shape creation.

**Comment by cvuosalo** [comment]
The comparison test differences are showing up in other PRs and are unrelated to this PR.

**Comment by cvuosalo** [comment]
+1

**Comment by silviodonato** [comment]
do you agree @ianna @civanch ?

**Comment by ianna** [comment]
@cvuosalo - IMHO it’s better to use type alias:
```c++
using DDBox = dd4hep::Box
```
A user can access its parameters as:
```c++
DDBox box;
auto x = box.x();
auto y = box.y();
auto z = box.z();
```

Here is an example of a `dd4hep::Box`

```c++

class Box : public Solid_type<TGeoBBox> {
  protected:
    /// Internal helper method to support object construction
    void make(const std::string& name, double x_val, double y_val, double z_val);

  public:
    /// Default constructor
    Box() = default;
    /// Move constructor
    Box(Box&& e) = default;
    /// Copy constructor
    Box(const Box& e) = default;
    /// Constructor to be used with an existing object
    template <typename Q> Box(const Q* p) : Solid_type<TGeoBBox>(p) { }
    /// Copy Constructor to be used with an existing object handle
    template <typename Q> Box(const Handle<Q>& e) : Solid_type<TGeoBBox>(e) { }

    /// Constructor to create an anonymous new box object (retrieves name from volume)
    Box(double x_val, double y_val, double z_val)
    { make("", x_val, y_val, z_val);    }
    /// Constructor to create a named new box object (retrieves name from volume)
    Box(const std::string& nam, double x_val, double y_val, double z_val)
    { make(nam.c_str(), x_val, y_val, z_val);    }

    /// Constructor to create an anonymous new box object (retrieves name from volume)
    template <typename X, typename Y, typename Z>
    Box(const X& x_val, const Y& y_val, const Z& z_val)
    { make("", _toDouble(x_val), _toDouble(y_val), _toDouble(z_val));    }
    /// Constructor to create a named new box object (retrieves name from volume)
    template <typename X, typename Y, typename Z>
    Box(const std::string& nam, const X& x_val, const Y& y_val, const Z& z_val)
    { make(nam.c_str(), _toDouble(x_val), _toDouble(y_val), _toDouble(z_val));  }

    /// Move Assignment operator
    Box& operator=(Box&& copy) = default;
    /// Copy Assignment operator
    Box& operator=(const Box& copy) = default;
    /// Set the box dimensions
    Box& setDimensions(double x_val, double y_val, double z_val);
    /// Access half "length" of the box
    double x() const;
    /// Access half "width" of the box
    double y() const;
    /// Access half "depth" of the box
    double z() const;
  };

```
If there is anything missing I’d ask @MarkusFrankATcernch to implement it

**Comment by ghugo83** [comment]
An issue is that, for many shapes, named parameter of interest are not accessible from DD4hep interface.

This is not the case for dd4hep::Box ( https://dd4hep.web.cern.ch/dd4hep/reference/classdd4hep_1_1Box.html ), which indeed has public member functions such as x(), y(), z().
But have a look at dd4hep::Tube for example ( https://dd4hep.web.cern.ch/dd4hep/reference/classdd4hep_1_1Tube.html ). There is no public member functions like rMin(), rMax(), dZ(), startPhi(), deltaPhi() ! The only way to access the parameters is to use dimensions(), which provides a vector of all the parameters. 

From what I understand, what is done in DetectorDescription/DDCMS/interface/DDShapes.h, is precisely to provide these missing functions 
(https://github.com/cms-sw/cmssw/blob/master/DetectorDescription/DDCMS/interface/DDShapes.h#L30 and https://github.com/cms-sw/cmssw/blob/master/DetectorDescription/DDCMS/src/DDShapes.cc#L115 ).

But definitely agree with @ianna here, these should be directly accessible from dd4hep interface. Why would we need to re-implement in CMSSW tub.rMax(), using a cast to a TGeo shape, to be able to access sth as trivial? Is it not sth pretty generic the other experiments do need as well anyway?

**Comment by MarkusFrankATcernch** [comment]
@ghugo83 
All dd4hep shapes are handles to the original TGeo shapes. In particular the operator-> is implemented, which gives you direct access to any member function of the TGeo shape.
Given:
`dd4hep::Box box = .....;`
Then 
`box.x()`  [The dd4hep call]
and
`box->GetDX()`  [The TGeo call]
are absolutely equivalent. 
Obviously the same mechanism holds for all other TGeo shapes (Cone, Tube, etc.) and the other objects.
It was a deliberate design decision to NOT mimic all possible accessors exposed by TGeo, but rather allow to access the underlying object. 
Enhancing implementations in the various handles were only implemented to achieve functionality, which a priori was not exposed by TGeo itself.

Cheers
Markus

**Comment by ghugo83** [comment]
@MarkusFrankATcernch Thanks a lot for your reply.
Ok makes sense now, so yep we need to go through the accessors from TGeo then.

File: DetectorDescription/DDCMS/interface/DDFilteredView.h [hunk 1]
```diff
@@ -179,6 +179,21 @@ namespace cms {
       return isA<dd4hep::Tube>(std::forward<Ts>(ts)...);
     }
 
+    template <typename... Ts>
+    auto isAPolycone(Ts&&... ts) const -> decltype(isA<dd4hep::Polycone>(std::forward<Ts>(ts)...)) {
```
Comment (ghugo83) [inline DetectorDescription/DDCMS/interface/DDFilteredView.h:183]: Touristic question here, didn't dig too much:
decltype(isA<dd4hep::Polycone>(std::forward<Ts>(ts)...))
Is that not... always a boolean ?

Comment (ghugo83) [reply DetectorDescription/DDCMS/interface/DDFilteredView.h:183]: template <typename... Ts>
bool isAPolycone(Ts&&... ts) const {
      return isA<dd4hep::Polycone>(std::forward<Ts>(ts)...);
}

Comment (MarkusFrankATcernch) [reply DetectorDescription/DDCMS/interface/DDFilteredView.h:183]: Not sure this needs the template:
```
bool isAPolycone(dd4hep::Solid solid)  {
  return dd4hep::isA<dd4hep::Polycone>(solid);
}
```
Should match all possible shapes as well.

**Comment by cvuosalo** [comment]
@ianna Yes, revision of `DDBox` and the older shapes is planned for a follow-up PR. I submitted an issue to remind us about it: #30964. I don't think polishing up the `DDShapes` is a high priority issue right now as we try to complete the migration.

**Comment by ianna** [comment]
> @ianna Yes, revision of `DDBox` and the older shapes is planned for a follow-up PR. I submitted an issue to remind us about it: #30964. I don't think polishing up the `DDShapes` is a high priority issue right now as we try to complete the migration.

@cvuosalo - it looks like we are pulling in different and conflicting directions - I'm trying to remove the very same code from CMSSW by migrating it to DD4hep and you are adding an extra layer :-)

I'm not a fan of ROOT naming style either, and a simpler solution would be to enhance the DD4hep handles.

BTW, CMS detector description uses the unnamed parameters for a reason. Firstly, it is almost impossible to decipher the shape coordinates in the code because it's either global or local depending on a parent and placement. Secondly, even if the names change from `dx1`, `dx2`, `dy1`, `dy2` to something `top` and `bottom` in `X` and `Y` how do you define the top and the bottom consistently?

**Comment by cvuosalo** [comment]
@ianna We need to come to agreement on the strategy for these shapes. It looks like there are three options:
* CMSSW DDShapes, as in this PR.
* Getting the DD4hep shapes like dd4hep::Cone, dd4hep::Polycone, etc. enhanced to provide named methods that return the lengths and angles of the shape. Note that I suggested this approach to Markus Frank, and his response was the basis for the implementation in this PR.
* Using unnamed parameters since it is difficult to find meaningful names for parameters whose meaning can change in different contexts.

**Comment by cvuosalo** [comment]
@ianna Could you please create an issue or draft PR concerning the code you are trying to move from CMSSW to DD4hep? It would be good to get a list so I and others know what code is likely to be changed or removed in the near future.

**Comment by MarkusFrankATcernch** [comment]
@cvuosalo  Carl, I can very well enhance the shapes in DD4hep. 
The question is: Where does one stop?
* If it is a set of very basic agreed quantities this is reasonable. 
* Things are getting unreasonable as soon as the entire TGeoShape interface is going to be exposed - or a significant part of it.
  Such an approach can only end up in a loose situation.

**Comment by cvuosalo** [comment]
hold

**Comment by cvuosalo** [comment]
We are reviewing shape designs, so I put this PR on hold.

**Comment by cvuosalo** [comment]
@ianna @MarkusFrankATcernch 
I have tried to review and consider all the uses of shape parameters. I favor a hybrid approach. For very simple cases, where only two or three parameters are needed and no shape is even mentioned, then the `parameters` method can be used. The geometry builders have examples like this.

In other cases, where the code is dealing with a specific shape and the parameters are somewhat complex, then the classes in `DDShapes.h` (in this PR) can be used. For example, extruded polygons are quite complex, and I would think dealing with a long list of numbered parameters could be confusing. The `DDExtrudedPolygon` class can be quite helpful in such a case.

The TGeo classes are not sufficient in themselves. The code in this PR is exactly the code that is needed to conveniently collect the values from the TGeo classes. Either we implement such interfaces once like in this PR, or else developers would have to implement their own interfaces to the TGeo classes over and over again.

Whether the functionality of the classes should be added to DD4hep is up to the DD4hep team. Markus has said he doesn't want to add this functionality to the DD4hep classes, and he favors this PR. I think that this PR should move forward, and we can continue to discuss what parts should or should not be moved into DD4hep.

What is your opinion? Can we move forward with this PR?

**Comment by cvuosalo** [comment]
unhold

**Comment by ianna** [comment]
> extruded polygon

@cvuosalo - IMHO, no, we should not go ahead with this PR. Merging this code will open a door for a much larger scale of the code refactoring - one should never allow users to depend on an incomplete solution: there should be one common API to the shapes.

Was this API enhancement requested by users? If yes, and as you've mentioned before this is work in progress, I'd suggest you keep it in a private branch and submit it to the DD4hep repository when it's ready and polished.

BTW, there is only one use case of `ExtrudedPolygon` in whole CMSSW:
https://cmssdt.cern.ch/lxr/source/Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalWafer8.cc#0054

`Polyhedra` is more widely used, for example: 
https://cmssdt.cern.ch/lxr/source/Geometry/HcalAlgo/plugins/dd4hep/DDHCalBarrelAlgo.cc#0354

Both shapes are used in the `HGCal` detector description, so they are performance critical.

**The main goal of the DD4hep migration is to reduce the detector description code base in CMSSW. If we cannot achieve that it's better to stick with a legacy DD rather than reimplementing it.** 

-1

**Comment by cvuosalo** [comment]
hold

**Comment by cvuosalo** [comment]
@ianna The current situation is that there are three unrelated interfaces for shapes. Progress on this problem will have to be incremental, like most software development. I doubt anyone is going to provide us a complete and comprehensive common shape API in one quantum leap. You and I have twice asked DD4hep to provide read access to shapes, and twice the answer has been no, so that avenue does not look promising right now. This PR, based upon Markus's code, was a modest attempt to add to the already existing DDShapes. If it cannot get approved, this path looks like a dead end. It seems there will be three unrelated interfaces for shapes in use for some time to come.

I will write a document to examine some possible options.

**Comment by ianna** [comment]
> @ianna The current situation is that there are three unrelated interfaces for shapes. Progress on this problem will have to be incremental, like most software development. I doubt anyone is going to provide us a complete and comprehensive common shape API in one quantum leap. You and I have twice asked DD4hep to provide read access to shapes, and twice the answer has been no, so that avenue does not look promising right now. This PR, based upon Markus's code, was a modest attempt to add to the already existing DDShapes. If it cannot get approved, this path looks like a dead end. It seems there will be three unrelated interfaces for shapes in use for some time to come.
> 
> I will write a document to examine some possible options.

Thanks. A set of user stories would help.

**Comment by civanch** [comment]
@ianna , may be we merge this PR and continue the discussion?

**Comment by ianna** [comment]
> @ianna , may be we merge this PR and continue the discussion?

I'd suggest we close it. The next version of DD4hep will provide the accessors needed for CMS. There is no need to multiply entities without necessity :-)

**Comment by ianna** [comment]
-1

**Comment by MarkusFrankATcernch** [comment]
@cvuosalo  @ianna 
You perhaps want to cross-check 
https://github.com/MarkusFrankATcernch/DD4hep/blob/master/DDCore/include/DD4hep/Shapes.h
before I make a merge request.

I added the basic data accessors to the shapes as suggested from Carl's implementation above.
We should have a common understanding, that we only add the basic data accessors to geometrical quantities.

**Comment by cvuosalo** [comment]
@MarkusFrankATcernch Thank you. Having these methods in DD4hep is the best solution.
The accessors look good, except what do we do about `TruncatedTube`? Does it lack `access()` methods? I didn't yet look into how to implement the accessor methods as wrappers for that class. For consistency, it would be nice to have accessors for `TruncatedTube`, but CMSSW currently only has one place where the parameters of a truncated tube are read, though that could change in the future.
Also, there is a typo or two in the comments, but I can point them out in the merge request.

**Comment by MarkusFrankATcernch** [comment]
@cvuosalo TruncatedTube is a bit difficult. 
The construction parameters of the TruncatedTube are stored as a string in the object's title (I know this is a hack), because I could not find an unambiguous solution to re-create the construction parameters by analyzing the composite shape.

**Comment by silviodonato** [comment]
@cvuosalo @ianna can we close this PR? If I understood correctly, the new features are going to be added in DD4HEP through https://github.com/AIDASoft/DD4hep/pull/697

**Comment by ianna** [comment]
@silviodonato - yes, I think we should close it.

**Comment by cvuosalo** [comment]
@silviodonato I was waiting for https://github.com/AIDASoft/DD4hep/pull/697 to be merged into DD4hep, but that process is taking a while. I can close this PR if it is causing confusion.
Some parts of this PR may return depending on the outcome of https://github.com/AIDASoft/DD4hep/pull/697.

**Pull request closed without merging**